v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package crd
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"time"
+
+	apiextensionsapi "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Cleaner struct {
+	apiExtensionsClient apiextensions.CustomResourceDefinitionInterface
+	client              client.Client
+	migrationBackoff    wait.Backoff
+	dryRun              bool
+	updated             int
+}
+
+func NewCleaner(apiExtensionsClient apiextensions.CustomResourceDefinitionInterface, client client.Client, dryRun bool) *Cleaner {
+	migrationBackoff := wait.Backoff{ // TODO: Still need to see if we want exponential backoff or linear is fine
+		Duration: 2 * time.Second, // wait 2s between attempts, this will help us in a state of conflict.
+		Steps:    3,               // 3 retry on error attempts per object
+	}
+	return &Cleaner{
+		apiExtensionsClient: apiExtensionsClient,
+		client:              client,
+		migrationBackoff:    migrationBackoff,
+		dryRun:              dryRun,
+	}
+}
+
+func (c *Cleaner) Run(ctx context.Context) error {
+	list, err := c.apiExtensionsClient.List(ctx, v1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if list == nil || len(list.Items) == 0 {
+		fmt.Println("found 0 results, make sure you have ASO CRDs installed")
+	}
+
+	crdRegexp := regexp.MustCompile(`.*\.azure\.com`)
+	deprecatedVersionRegexp := regexp.MustCompile(`v1alpha1api\d{8}(preview)?(storage)?`)
+	for _, crd := range list.Items {
+		crd := crd
+
+		if !crdRegexp.MatchString(crd.Name) {
+			continue
+		}
+
+		newStoredVersions := removeMatchingStoredVersions(crd.Status.StoredVersions, deprecatedVersionRegexp)
+		objectsToMigrate, err := c.getObjectsForMigration(ctx, crd, deprecatedVersionRegexp)
+		if err != nil {
+			return err
+		}
+
+		if len(newStoredVersions) > 0 && len(newStoredVersions) != len(crd.Status.StoredVersions) {
+
+			err = c.migrateObjects(ctx, objectsToMigrate)
+			if err != nil {
+				return err
+			}
+
+			err = c.updateStorageVersions(ctx, crd, newStoredVersions)
+			if err != nil {
+				return err
+			}
+
+			c.updated++
+		} else {
+			fmt.Printf("nothing to update for '%s'\n", crd.Name)
+		}
+	}
+
+	if !c.dryRun {
+		fmt.Printf("updated %d CRD(s)\n", c.updated)
+
+	}
+
+	return nil
+}
+
+func (c *Cleaner) updateStorageVersions(
+	ctx context.Context,
+	crd apiextensionsapi.CustomResourceDefinition,
+	newStoredVersions []string) error {
+
+	if c.dryRun {
```

Comment:
minor: On all of these functions, instead of this `if/else` pattern, maybe consider:

```
// Log
if c.dryRun {
    // return
}


```

I think that's a bit cleaner?